From b3b992c167918fe966d07c2adcb95f6efbdbac5e Mon Sep 17 00:00:00 2001
From: Matt Riedemann <mriedem.os@gmail.com>
Date: Thu, 30 Aug 2018 14:19:02 +0000
Subject: Revert "Update resources once in update_available_resource"

This reverts commit c9b74bcfa09d11c2046ce1bfb6dd8463b3a2f3b0.

This is causing the ResourceTracker to report 0.0 allocation
ratio values to Placement inventory during the
update_available_resource periodic task because
_init_compute_node calls _copy_resources which copies over
the default allocation ratio values from config (default
to 0.0) but then we don't "fix" the 0.0 allocation ratios
via ComputeNode.save() (where ComputeNode._from_db_object
would actually hard-code the appropriate non-0.0 values)
before passing the 0.0 allocation ratios to Placement, which
results in NoValidHost errors during scheduling since
Placement doesn't think there are any valid allocation
candidates available.

Change-Id: Iba24d3e735a274d46fe3608ba9d3b7fdb023dfd4
Closes-Bug: #1789654
---
 nova/compute/resource_tracker.py                 |  3 +++
 nova/tests/unit/compute/test_resource_tracker.py | 14 +++-----------
 2 files changed, 6 insertions(+), 11 deletions(-)

diff --git a/nova/compute/resource_tracker.py b/nova/compute/resource_tracker.py
index 71a743b..ac818a8 100644
--- a/nova/compute/resource_tracker.py
+++ b/nova/compute/resource_tracker.py
@@ -561,6 +561,7 @@ class ResourceTracker(object):
             cn = self.compute_nodes[nodename]
             self._copy_resources(cn, resources)
             self._setup_pci_tracker(context, cn, resources)
+            self._update(context, cn)
             return
 
         # now try to get the compute node record from the
@@ -570,6 +571,7 @@ class ResourceTracker(object):
             self.compute_nodes[nodename] = cn
             self._copy_resources(cn, resources)
             self._setup_pci_tracker(context, cn, resources)
+            self._update(context, cn)
             return
 
         if self._check_for_nodes_rebalance(context, resources, nodename):
@@ -588,6 +590,7 @@ class ResourceTracker(object):
                  {'host': self.host, 'node': nodename, 'uuid': cn.uuid})
 
         self._setup_pci_tracker(context, cn, resources)
+        self._update(context, cn)
 
     def _setup_pci_tracker(self, context, compute_node, resources):
         if not self.pci_tracker:
diff --git a/nova/tests/unit/compute/test_resource_tracker.py b/nova/tests/unit/compute/test_resource_tracker.py
index 0ec2b15..959e6af 100644
--- a/nova/tests/unit/compute/test_resource_tracker.py
+++ b/nova/tests/unit/compute/test_resource_tracker.py
@@ -573,7 +573,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
                 return_value=objects.InstancePCIRequests(requests=[]))
@@ -614,7 +613,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.compute.utils.is_volume_backed_instance')
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
@@ -664,7 +662,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
                 return_value=objects.InstancePCIRequests(requests=[]))
@@ -728,7 +725,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
                 return_value=objects.InstancePCIRequests(requests=[]))
@@ -791,7 +787,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
                 return_value=objects.InstancePCIRequests(requests=[]))
@@ -851,7 +846,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
                 return_value=objects.InstancePCIRequests(requests=[]))
@@ -908,7 +902,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
     @mock.patch('nova.compute.utils.is_volume_backed_instance')
     @mock.patch('nova.objects.InstancePCIRequests.get_by_instance',
@@ -980,7 +973,6 @@ class TestUpdateAvailableResources(BaseTestCase):
         actual_resources = update_mock.call_args[0][1]
         self.assertTrue(obj_base.obj_equal_prims(expected_resources,
                                                  actual_resources))
-        update_mock.assert_called_once()
 
 
 class TestInitComputeNode(BaseTestCase):
@@ -1006,7 +998,7 @@ class TestInitComputeNode(BaseTestCase):
         self.assertFalse(get_mock.called)
         self.assertFalse(create_mock.called)
         self.assertTrue(pci_mock.called)
-        self.assertFalse(update_mock.called)
+        self.assertTrue(update_mock.called)
 
     @mock.patch('nova.objects.PciDeviceList.get_by_compute_node',
                 return_value=objects.PciDeviceList())
@@ -1030,7 +1022,7 @@ class TestInitComputeNode(BaseTestCase):
         get_mock.assert_called_once_with(mock.sentinel.ctx, _HOSTNAME,
                                          _NODENAME)
         self.assertFalse(create_mock.called)
-        self.assertFalse(update_mock.called)
+        self.assertTrue(update_mock.called)
 
     @mock.patch('nova.objects.ComputeNodeList.get_by_hypervisor')
     @mock.patch('nova.objects.PciDeviceList.get_by_compute_node',
@@ -1195,7 +1187,7 @@ class TestInitComputeNode(BaseTestCase):
         self.assertTrue(obj_base.obj_equal_prims(expected_compute, cn))
         pci_tracker_mock.assert_called_once_with(mock.sentinel.ctx,
                                                  42)
-        self.assertFalse(update_mock.called)
+        self.assertTrue(update_mock.called)
 
     @mock.patch('nova.objects.ComputeNodeList.get_by_hypervisor')
     @mock.patch('nova.objects.PciDeviceList.get_by_compute_node',
-- 
cgit v1.1

